lib/policies/speculative-execution.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+var util = require('util');
+var errors = require('../errors');
+
+/** @module policies/speculativeExecution */
+
+/**
+ * @classdesc
+ * The policy that decides if the driver will send speculative queries to the next hosts when the current host takes too
+ * long to respond.
+ * <p>Note that only idempotent statements will be speculatively retried.</p>
+ * @constructor
+ * @abstract
+ */
+function SpeculativeExecutionPolicy() {
+  
+}
+
+/**
+ * Initialization method that gets invoked on Client startup.
+ * @param {Client} client
+ * @abstract
+ */
+SpeculativeExecutionPolicy.prototype.init = function (client) {
+
+};
+
+/**
+ * Gets invoked at client shutdown, giving the opportunity to the implementor to perform cleanup.
+ * @abstract
+ */
+SpeculativeExecutionPolicy.prototype.shutdown = function () {
+
+};
+
+/**
+ * Gets the plan to use for a new query.
+ * Returns an object with a <code>nextExecution()</code> method, which returns a positive number representing the
+ * amount of milliseconds to delay the next execution or a non-negative number to avoid further executions.
+ * @param {String} keyspace The currently logged keyspace.
+ * @param {String|Array<String>} queryInfo The query, or queries in the case of batches, for which to build a plan.
+ * @return {{nextExecution: function}}
+ * @abstract
+ */
+SpeculativeExecutionPolicy.prototype.newPlan = function (keyspace, queryInfo) {
+  throw new Error('You must implement newPlan() method in the SpeculativeExecutionPolicy');
+};
+
+/**
+ * Creates a new instance of NoSpeculativeExecutionPolicy.
+ * @classdesc
+ * A {@link SpeculativeExecutionPolicy} that never schedules speculative executions.
+ * @constructor
+ * @extends {SpeculativeExecutionPolicy}
+ */
+function NoSpeculativeExecutionPolicy() {
+  this._plan = {
+    nextExecution: function () {
+      return -1;
+    }
+  };
+}
+
+util.inherits(NoSpeculativeExecutionPolicy, SpeculativeExecutionPolicy);
+
+NoSpeculativeExecutionPolicy.prototype.newPlan = function () {
+  return this._plan;
+};
+
+
+/**
+ * Creates a new instance of ConstantSpeculativeExecutionPolicy.
+ * @classdesc
+ * A {@link SpeculativeExecutionPolicy} that schedules a given number of speculative executions,
+ * separated by a fixed delay.
+ * @constructor
+ * @param {Number} delay The delay between each speculative execution. Must be strictly positive.
+ * @param {Number} maxSpeculativeExecutions The number of speculative executions. Must be strictly positive.
+ * @extends {SpeculativeExecutionPolicy}
+ */
+function ConstantSpeculativeExecutionPolicy(delay, maxSpeculativeExecutions) {
+  if (!(delay > 0)) {
```

Comment:
One upcoming change in the java driver that would be nice to carry along here is to allow 0 ms delay between speculative executions ([JAVA-1490](https://github.com/datastax/java-driver/pull/856)).  For users who need very quick response SLAs (i.e. 5ms), the ability to send multiple requests out immediately helps quite a bit.  The other nice part about this is you could special case 0ms delays by sending the requests immediately rather than scheduling them on a timer (which may have a perceived delay as well).